### PR TITLE
Add schema files for request validation.

### PR DESCRIPTION
**Pull request by jbasden**
This commit adds the JSON schema for log_events and metric_datum.
It also adds the core logic for schema.py.

### CODE REVIEW

File: metricpublisher/schemas/log_event.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Schema for the log_event lambda request object",
+    "type": "object",
+    "properties": {
+        "request_id": {
+            "type": "number"
```
Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: request_id should be a string

File: metricpublisher/schemas/metric_datum.json [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+                "Terabits/Second",
+                "Count/Second",
+                "None"
+            ]
+        },
+        "StorageResolution": {
+            "type": "number"
+        }
+    },
+    "required": [
```
Comment (ylynn) [inline metricpublisher/schemas/metric_datum.json]: Did you try these two cases with cli: 1. both Value and StatisticValues exist 2. both Value and StatisticValues don't exist?

**Review by ylynn — CHANGES_REQUESTED** [review]

File: metricpublisher/schemas/log_event.json (hunk 1 above)
Comment (jbasden) [reply metricpublisher/schemas/log_event.json]: Made the change.

File: metricpublisher/schemas/metric_datum.json (hunk 2 above)
Comment (jbasden) [reply metricpublisher/schemas/metric_datum.json]: Yes - both of them give an error. I was unaware that they were supposed to be collectively exhaustive and mutually exclusive. Should this be reflected in the JSON schema?

Comment (ylynn) [reply metricpublisher/schemas/metric_datum.json]: Yes.

**Review by ylynn — CHANGES_REQUESTED** [review]
Include unit tests for validate function, otherwise you cannot know if they will work.

**Comment by ylynn** [comment]
Describe what test you ran?

**Comment by ylynn** [comment]
I was asking you to describe what test you ran in your pull request description. Like I said before, you can just say "run make to test" or "make passed".

**Review by ylynn — CHANGES_REQUESTED** [review]
Add tests for invalid "Dimension" value.

File: metricpublisher/schemas/log_event.json [hunk 3]
```diff
@@ -0,0 +1,129 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Schema for the log_event lambda request object",
+    "type": "object",
+    "properties": {
+        "request_id": {
+            "type": "string"
+        },
+        "metric_data": {
```
Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: For the attributes that have properties under it, have definition files for them separately and then ref them. See example https://github.com/awslabs/aws-serverless-dynamodb-event-sourcing/tree/master/app/schemas

Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: For the attributes that have properties under it, have definition files for them separately and then ref them. See example https://github.com/awslabs/aws-serverless-dynamodb-event-sourcing/tree/master/app/schemas

File: metricpublisher/schemas/log_event.json [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            "type": "array",
+            "items": {
+                "type": "object",
+                "properties": {
+                    "MetricName": {
+                        "type": "string"
+                    },
+                    "Dimensions": {
+                        "type": "array",
+                        "properties": {
```
Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: This is wrong. For type array, you should have something like this: 
```
"type": "array",
  "items": {
    "type": "object",
    "properties": {
```
So the other problem is that your tests did not help catch this.

**Review by ylynn — CHANGES_REQUESTED** [review]

File: test/unit/events.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+                    "Maximum": 12.17
+                },
+                "Unit": "Seconds",
+                "StorageResolution": "12"
+            }
+        ]
+    }
+
+@pytest.fixture()
+def Dimension_type_invalid():
```
Comment (ylynn) [inline test/unit/events.py]: Dimensions invalid cases include: 
1. not an array
2. wrong type of item in the array
3. wrong property of item

**Review by ylynn — CHANGES_REQUESTED** [review]

File: test/unit/events.py (hunk 5 above)
Comment (jbasden) [reply test/unit/events.py]: Dimension_type_invalid deals with 1. Will add separate tests for 2 and 3.

File: metricpublisher/schemas/log_event.json [hunk 6]
```diff
@@ -0,0 +1,162 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "title": "Schema for the log_event lambda request object",
+    "type": "object",
+    "definitions": {
+        "request_id": {
```
Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: Only include the attributes have "items" or "properties" under them in "definitions". Simple ones like "request_id" does not need to have separate definitions.

**Review by ylynn — CHANGES_REQUESTED** [review]

File: metricpublisher/schemas/log_event.json (hunk 6 above)
Comment (jbasden) [reply metricpublisher/schemas/log_event.json]: Oh got ya. I thought you meant to have separate definitions for the properties themselves, not the ones that have properties underneath them...

File: metricpublisher/schemas/log_event.json [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+                "Value": {
+                    "type": "string"
+                }
+            },
+            "required": [
+                "Name",
+                "Value"
+            ]
+        },
+        "Dimensions": {
```
Comment (ylynn) [inline metricpublisher/schemas/log_event.json]: You have some attributes capitalized ("Dimensions" "StatisticValues" "SampleCount" ...) and some not ("metric_datum" "metric_data"). Make them all lower case to be consistent.

**Review by ylynn — CHANGES_REQUESTED** [review]

**Review by ylynn — APPROVED** [review]

File: metricpublisher/lambda_handler.py [hunk 8]
```diff
@@ -1 +1,35 @@
...
+
+from jsonschema import ValidationError
+import schema
+
+
+def log_event(event, context):
+    """Log event.
+
+    Parameters:
+        event (dict): the user input
```
Comment (honglu) [inline metricpublisher/lambda_handler.py:11]: "the user input" is not very helpful in the doc. Better to more descriptive by mentioning it is the metric information user would like to emit.

**Review by honglu — APPROVED** [review]

**Pull request merged**